File: src/main/java/de/hinundhergestellt/jhuh/vendors/shopify/ShopifyProductClient.java
package de.hinundhergestellt.jhuh.vendors.shopify;

import com.netflix.graphql.dgs.client.GraphQLClient;
import com.netflix.graphql.dgs.client.codegen.GraphQLQueryRequest;
import com.shopify.admin.client.ProductCreateGraphQLQuery;
import com.shopify.admin.client.ProductCreateProjectionRoot;
import com.shopify.admin.client.ProductsGraphQLQuery;
import com.shopify.admin.client.ProductsProjectionRoot;
import com.shopify.admin.types.PageInfo;
import com.shopify.admin.types.Product;
import com.shopify.admin.types.ProductConnection;
import com.shopify.admin.types.ProductCreateInput;
import com.shopify.admin.types.ProductCreatePayload;
import com.shopify.admin.types.ProductEdge;
import org.apache.commons.lang3.tuple.Pair;
import org.springframework.stereotype.Component;

import java.util.stream.Stream;

@Component
public class ShopifyProductClient {

    private final GraphQLClient apiClient;

    public ShopifyProductClient(GraphQLClient shopifyApiClient) {
        this.apiClient = shopifyApiClient;
    }

    public Stream<ShopifyProduct> findAll() {
        return PagingIterator.stream(this::findAll);
    }

    public void save(Product product) {
        if (product.getId() != null) {
            throw new IllegalArgumentException("Product id must be null");
        }

        var productInput = new ProductCreateInput();
        productInput.setTitle(product.getTitle());
        productInput.setVendor(product.getVendor());
        productInput.setProductType(product.getProductType());
        productInput.setTags(product.getTags());

        var query = ProductCreateGraphQLQuery.newRequest()
                .product(productInput)
                .build();

        // @formatter:off
        var root = new ProductCreateProjectionRoot<>()
                .product()
                    .id()
                    .parent()
                .userErrors()
                    .message()
                    .field();
        // @formatter:on

        var request = new GraphQLQueryRequest(query, root);
        var response = apiClient.executeQuery(request.serialize());
        var payload = response.extractValueAsObject("productCreate", ProductCreatePayload.class);
        if (!payload.getUserErrors().isEmpty()) {
            throw new IllegalArgumentException("Product creation failed: " + payload.getUserErrors());
        }
        product.setId(payload.getProduct().getId());
    }

    private Pair<Stream<ShopifyProduct>, PageInfo> findAll(String after) {
        var query = ProductsGraphQLQuery.newRequest()
                .first(100)
                .after(after)
                .build();

        // @formatter:off
        var root = new ProductsProjectionRoot<>()
                .edges()
                    .node()
                        .handle()
                        .id()
                        .title()
                        .vendor()
                        .productType()
                        .tags()
                        .variants(100, null, null, null, null, null)
                            .edges()
                                .node()
                                    .id()
                                    .title()
                                    .price()
                                    .sku()
                                    .barcode()
                                    //.inventoryItem().id().parent()
                                    .parent()
                                .parent()
                            .pageInfo()
                                .hasNextPage()
                                .parent()
                            .parent()
                        .parent()
                    .parent()
                .pageInfo()
                    .hasNextPage()
                    .endCursor();
        // @formatter:on

        var request = new GraphQLQueryRequest(query, root);
        var response = apiClient.executeQuery(request.serialize());
        if (response.hasErrors()) {
            throw new RuntimeException(response.getErrors().toString()); // TODO
        }
        var payload = response.extractValueAsObject("products", ProductConnection.class);
        return Pair.ofNonNull(
                payload.getEdges().stream()
                        .map(ProductEdge::getNode)
                        .map(ShopifyProduct::new),
                payload.getPageInfo()
        );
    }
}

File: src/main/java/de/hinundhergestellt/jhuh/vendors/shopify/ShopifyProductClient.kt
package de.hinundhergestellt.jhuh.vendors.shopify

import com.netflix.graphql.dgs.client.GraphQLClient
import com.netflix.graphql.dgs.client.codegen.BaseSubProjectionNode
import com.netflix.graphql.dgs.client.codegen.GraphQLQueryRequest
import com.shopify.admin.client.ProductCreateGraphQLQuery
import com.shopify.admin.client.ProductCreateProjectionRoot
import com.shopify.admin.client.ProductDeleteGraphQLQuery
import com.shopify.admin.client.ProductDeleteProjectionRoot
import com.shopify.admin.client.ProductsGraphQLQuery
import com.shopify.admin.client.ProductsProjectionRoot
import com.shopify.admin.types.*
import org.springframework.stereotype.Component

@Component
class ShopifyProductClient(
    private val apiClient: GraphQLClient
) {
    fun findAll() = pageAll {
        findAll(it)
    }

    fun create(product: ShopifyProduct) {
        require(product.id == null) { "Product id must be null" }

        val query = ProductCreateGraphQLQuery.newRequest()
            .product(product.toProductCreateInput())
            .build()

        // @formatter:off
        val root = ProductCreateProjectionRoot<BaseSubProjectionNode<*, *>?, BaseSubProjectionNode<*, *>?>()
            .product()
                .id()
                .options()
                    .id()
                    .parent()
                .parent()
            .userErrors()
                .message()
                .field()
        // @formatter:on

        val request = GraphQLQueryRequest(query, root)
        val response = apiClient.executeQuery(request.serialize())
        val payload = response.extractValueAsObject("productCreate", ProductCreatePayload::class.java)
        require(payload.userErrors.isEmpty()) { "Product creation failed: " + payload.userErrors }

        product.options.zip(payload.product.options).forEach { (option, created) -> option.id = created.id!! }
        product.id = payload.product.id!!
    }

    fun delete(product: ShopifyProduct) {
        val input = ProductDeleteInput()
        input.id = product.id

        val query = ProductDeleteGraphQLQuery.newRequest()
            .input(input)
            .build()

        // @formatter:off
        val root = ProductDeleteProjectionRoot<BaseSubProjectionNode<*, *>, BaseSubProjectionNode<*, *>>()
            .userErrors()
                .message()
                .field()
        // @formatter:on

        val request = GraphQLQueryRequest(query, root)
        val response = apiClient.executeQuery(request.serialize())
        require(!response.hasErrors()) { "Product delete failed: " + response.errors }
        val payload = response.extractValueAsObject("productDelete", ProductDeletePayload::class.java)
        require(payload.userErrors.isEmpty()) { "Product delete failed: " + payload.userErrors }
    }

    private fun findAll(after: String?): Pair<Sequence<ShopifyProduct>, PageInfo> {
        val query = ProductsGraphQLQuery.newRequest()
            .first(250)
            .after(after)
            .build()

        // @formatter:off
        val root = ProductsProjectionRoot<BaseSubProjectionNode<*, *>?, BaseSubProjectionNode<*, *>?>()
            .edges()
                .node()
                    .handle()
                    .id()
                    .title()
                    .vendor()
                    .productType()
                    .tags()
                    .variants(250, null, null, null, null, null)
                        .edges()
                            .node()
                                .id()
                                .title()
                                .price()
                                .sku()
                                .barcode()
                                .selectedOptions()
                                    .name()
                                    .value()
                                    .parent()
                                .parent()
                            .parent()
                        .pageInfo()
                            .hasNextPage()
                            .parent()
                        .parent()
                    .options()
                        .name()
                        .parent()
                    .parent()
                .parent()
            .pageInfo()
                .hasNextPage()
                .endCursor()
        // @formatter:on

        val request = GraphQLQueryRequest(query, root)
        val response = apiClient.executeQuery(request.serialize())
        require(!response.hasErrors()) { "Product collections find failed: " + response.errors }
        val payload = response.extractValueAsObject("products", ProductConnection::class.java)
        return Pair(
            payload.edges.asSequence()
                .map { it.node }
                .map { ShopifyProduct(it) },
            payload.pageInfo
        )
    }
}
